FLT/Deformations/Representable.lean

```diff
@@ -17,4 +20,82 @@ lemma isCorepresentable_deformationFunctor :
     (deformationFunctor n G 𝓞 ρ).toFunctor.IsCorepresentable := by
   sorry -- de Smit and Lenstra, Proposition 2.3 (1).
 
+end
+
+noncomputable section
+
+local notation3 "Γ" K:max => Field.absoluteGaloisGroup K
+local notation3 K:max "ᵃˡᵍ" => AlgebraicClosure K
+local notation "Ω" K => IsDedekindDomain.HeightOneSpectrum (NumberField.RingOfIntegers K)
+
+/-!
+Suppose `K` is a totally real number field with even degree, `l > 3` is a prime,
+`𝓞` an `ℤₗ`-algebra which is noetherian complete local with finite residue field.
+-/
+
+variable (𝓞 : Type u) [CommRing 𝓞] [IsLocalRing 𝓞] [IsNoetherianRing 𝓞]
+  [Finite (ResidueField 𝓞)] [IsAdicComplete (maximalIdeal 𝓞) 𝓞]
+variable {K : Type u} [Field K] [NumberField K]
+variable [NumberField.IsTotallyReal K] (hK : Even (Module.finrank ℚ K))
+variable (l : ℕ) [Fact l.Prime] (hl : 3 < l) [Algebra ℤ_[l] 𝓞]
+
+/-!
+Let `S` be a finite sets of finite primes of `K` not dividing `l`,
+and `ρ : Gₖ → GL₂(κ)` (where `κ` is the residue field of `𝓞`)
+be an absolutely irreducible galois representation.
+-/
+
+variable (S : Finset (Ω K)) (hS : ∀ v : Ω K, ↑l ∈ v.asIdeal → v ∉ S)
+variable (ρ : (repnFunctor (Fin 2) (Γ K) 𝓞).obj .residueField)
+variable [(toRepresentation ρ).IsAbsolutelyIrreducible]
+
+/--
+The functor of `S`-lifts of `ρ`, consisting of representations of `GL₂` satisfying:
+1. It is a lift of `ρ`.
+2. Its determinant is equal to `εₗ`, the `l`-adic cyclotomic character.
+3. It is unramified outside `l` and `S`.
+4. Its trace is `2` on `ker(Iᵥ → k(v)ˣ)`.
+5. It is flat at all `v | l`.
+-/
+def SLiftFunctor : Subfunctor (repnFunctor (Fin 2) (Γ K) 𝓞) :=
+  liftFunctor (Fin 2) _ 𝓞 ρ ⊓
+  detConditionFunctor (Fin 2) 𝓞 l ⊓
+  (⨅ (v : Ω K) (_ : ↑l ∉ v.asIdeal) (_ : v ∉ S), unramifiedFunctor (Fin 2) 𝓞 v) ⊓
+  (⨅ (v : Ω K) (_ : v ∈ S), traceConditionFunctor 𝓞 v) ⊓
+  (⨅ (v : Ω K) (_ : ↑l ∉ v.asIdeal), flatFunctor (Fin 2) 𝓞 v)
+
+/--
+The functor of narrow `S`-lifts of `ρ`, consisting of representations of `GL₂` satisfying:
+1. It is a lift of `ρ`.
+2. Its determinant is equal to `εₗ`, the `l`-adic cyclotomic character.
+3. It is unramified outside `l` and `S`.
+4. Its trace is `2` on `Iᵥ`.
+5. It is flat at all `v | l`.
+-/
+def narrowSLiftFunctor : Subfunctor (repnFunctor (Fin 2) (Γ K) 𝓞) :=
+  liftFunctor (Fin 2) _ 𝓞 ρ ⊓
+  detConditionFunctor (Fin 2) 𝓞 l ⊓
+  (⨅ (v : Ω K) (_ : ↑l ∉ v.asIdeal) (_ : v ∉ S), unramifiedFunctor (Fin 2) 𝓞 v) ⊓
+  (⨅ (v : Ω K) (_ : v ∈ S), narrowTraceConditionFunctor 𝓞 v) ⊓
+  (⨅ (v : Ω K) (_ : ↑l ∉ v.asIdeal), flatFunctor (Fin 2) 𝓞 v)
+
+variable (hρ : ρ ∈ (narrowSLiftFunctor 𝓞 l S ρ).obj _)
+
+include hK hl hS hρ in
+lemma isCorepresentable_narrowSLiftFunctor :
+    (narrowSLiftFunctor 𝓞 l S ρ).toFunctor.IsCorepresentable := by
+  sorry
```

Comment:
Do we have a strat for this?

Reply:
I have not yet spent much time on this. 
[Böckle's notes](https://typo.iwr.uni-heidelberg.de/fileadmin/groups/arithgeo/templates/data/Gebhard_Boeckle/Boeckle-Barcelona-DefTheory-Rev.pdf) seems to have an explicit construction of some sorts (Proposition 1.3.1 (a)) which we can probably then cut out the quotient that we need. But I have not yet made sure that all conditions are met, nor if there exists a better reference.